core/embed/rust/src/ui/display/font.rs

```diff
@@ -106,7 +103,106 @@ impl Glyph {
+/// });
+/// ```
+pub struct GlyphData {
+    font: Font,
+    #[cfg(feature = "translations")]
+    translations_guard: Option<RwLockReadGuard<'static, Option<Translations<'static>>>>,
+}
+
+impl GlyphData {
+    fn new(font: Font) -> Self {
+        #[cfg(feature = "translations")]
+        let translations_guard = flash::get().ok();
+
+        Self {
+            font,
+            #[cfg(feature = "translations")]
+            translations_guard,
+        }
+    }
+
+    pub fn get_glyph(&self, ch: char) -> Glyph<'_> {
+        let ch = match ch {
+            '\u{00a0}' => '\u{0020}',
+            c => c,
+        };
+        let gl_data = self.get_glyph_data(ch as u16);
+
+        ensure!(gl_data.is_some(), "Failed to load glyph");
+        Glyph::load(gl_data.unwrap())
+    }
+
+    fn calculate_glyph_size(header: &[u8]) -> usize {
+        let width = header[0] as i16;
+        let height = header[1] as i16;
+
+        let data_bytes = match constant::FONT_BPP {
+            1 => (width * height + 7) / 8, // packed bits
+            2 => (width * height + 3) / 4, // packed bits
+            4 => (width + 1) / 2 * height, // row aligned to bytes
+            8 => width * height,
+            _ => fatal_error!("Unsupported font bpp"),
+        };
+
+        5 + data_bytes as usize // header (5 bytes) + bitmap data
+    }
+
+    fn get_glyph_data(&self, codepoint: u16) -> Option<&[u8]> {
+        display::get_font_info(self.font.into()).map(|font_info| {
+            if codepoint >= ' ' as u16 && codepoint < 0x7F {
+                // ASCII character
+                unsafe {
+                    let ptr = *font_info
+                        .glyph_data
+                        .offset((codepoint - ' ' as u16) as isize);
+                    let header = slice::from_raw_parts(ptr, 2);
+                    let full_size = Self::calculate_glyph_size(header);
+                    slice::from_raw_parts(ptr, full_size)
+                }
+            } else {
+                #[cfg(feature = "translations")]
+                {
+                    if codepoint >= 0x7F {
+                        // UTF8 character from embedded blob
+                        if let Some(guard) = &self.translations_guard {
+                            if let Some(translations) = guard.as_ref() {
+                                if let Some(glyph) =
+                                    translations.get_utf8_glyph(codepoint, self.font as u16)
+                                {
+                                    return glyph;
+                                }
+                            }
+                        }
+                    }
+                }
+                unsafe {
+                    let ptr = font_info.glyph_nonprintable;
+                    let header = slice::from_raw_parts(ptr, 2);
+                    let full_size = Self::calculate_glyph_size(header);
+                    slice::from_raw_parts(ptr, full_size)
+                }
```

Comment:
this should be a its own `unsafe fn` (we could even go `Glyph::from_slice` and `unsafe fn Glyph::from_ptr`, but don't have to do that here)
AND there should be `pub fn GlyphData::nonprintable(&self) -> Glyph` that just calls this unsafe fn on `font_info.glyph_nonprintable`